### PR TITLE
feat(modals): support stacking modals, remove bootstrap modals dependency

### LINKED ISSUE #3516: [A11Y] Add missing attributes to modal
**Note:** this could be performed during a migration to `<dialog>` HTML elements.

### PR DESCRIPTION
**Pull request by jaspervriends**
Fixes #3516

**Changes proposed in this pull request:**
In this PR I've worked on making it possible to have multiple dialogs opened at the same time and make them stackable. I have partly used code of #3246 and altered it. It also removed the bootstrap modal dependency.

However, I did not proceed using the native HTML Dialog element for this PR due to lack of animations (for example, the backdrop is not animatable) and responsiveness with the interface.

A third parameter has been added to the `app.modal.show` function which is a boolean and defaults to `false`. This means that all already existing modals will keep their intended behaviour.

I also copied the new 'dismissable' options from the other PR, such as an option to hide the dialog on escape key, clicking on the backdrop or using the close button and managing them seperately per modal. This means there is a deprecation too which was mentioned in the code.

**Reviewers should focus on:**
- Current modals are opening and closing as intended
- Stacking multiple modals works
- Responsiveness is still good

**Necessity**

- [x] Has the problem that is being solved here been clearly explained?
- [x] If applicable, have various options for solving this problem been considered?
- [x] For core PRs, does this need to be in core, or could it be in an extension?
- [x] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [ ] Backend changes: tests are green (run `composer test`).
- [x] Core developer confirmed locally this works as intended.
- [x] Tests have been added, or are not appropriate here.

### CODE REVIEW

File: framework/core/js/src/common/components/ModalManager.tsx
Comment (davwheat) [inline framework/core/js/src/common/components/ModalManager.tsx]: Is this needed? Elements with the same zIndex should stack as per normal DOM hierarchy rules, where the last element is on top.

If we are going to do it this way, we should use the CSS variable for modal z-index and either:

1. Provide an custom attribute, such as `--modal-number`, then use `calc()` to set the z-index

2. Fetch the value of the modal z-index attribute on the `::root` selector and use that instead of a hardcoded value.

I expect number 1 will be easier.

File: framework/core/js/src/common/components/ModalManager.tsx
Comment (davwheat) [inline framework/core/js/src/common/components/ModalManager.tsx]: I would prefer to use `data-modal-key`.

File: framework/core/js/src/common/components/ModalManager.tsx
Comment (davwheat) [inline framework/core/js/src/common/components/ModalManager.tsx]: Could we rename this to more suitable match the doc block?


```suggestion
  private get activeDialogElement(): HTMLElement {
```

File: framework/core/js/src/common/components/ModalManager.tsx
Comment (davwheat) [inline framework/core/js/src/common/components/ModalManager.tsx]:
```suggestion
  private get activeBackdropElement(): HTMLElement {
```

File: framework/core/js/src/common/states/ModalManagerState.ts
Comment (davwheat) [inline framework/core/js/src/common/states/ModalManagerState.ts]: We should use `boolean` primitive rather than the `Boolean` object.

```suggestion
  show(componentClass: UnsafeModalClass, attrs: Record<string, unknown> = {}, stackModal: boolean = false): void {
```

File: framework/core/js/src/common/components/ModalManager.tsx
Comment (davwheat) [inline framework/core/js/src/common/components/ModalManager.tsx]: We should also add `aria-hidden="true"` to all models except the active one.

Also `role="dialog"`.

**Review by davwheat — COMMENTED** [review]
Just some comments so far. Haven't had time to properly review yet.

**Comment by jaspervriends** [comment]
Thanks for the feedback, will work on it later this week

**Comment by jaspervriends** [comment]
Processed feedback :)

**Review by davwheat — DISMISSED** [review]
I've addressed some things with commits on this PR, but would like a sanity check for my changes from another core dev before looking to merge this.

Thank you for your work on this! This is a big improvement for Flarum's modals system! :)

File: framework/core/js/src/forum/components/LogInModal.tsx
Comment (SychO9) [inline framework/core/js/src/forum/components/LogInModal.tsx]: I don't believe this should be stackable 🤔

File: framework/core/less/common/Modal.less
Comment (SychO9) [inline framework/core/less/common/Modal.less]: `.modal-open` definition is dead code now I think

File: framework/core/js/src/common/states/ModalManagerState.ts
Comment (SychO9) [inline framework/core/js/src/common/states/ModalManagerState.ts]: in 2.0 we'll probably keep the new `stackModal` argument set to `false` by default (or `true`). So I think we could dro this signature and the deprecation mark. Thoughts?

**Review by SychO9 — DISMISSED** [review]
Code looks good! a few comments and a bug I noticed. Thanks a lot for the PR!

There's an issue with the scrolling behavior, I think a video will better explain it:
https://user-images.githubusercontent.com/20267363/178266180-69aaa829-b528-4808-999f-4693bd15b3a5.mp4

File: framework/core/js/src/forum/components/LogInModal.tsx
Comment (davwheat) [reply framework/core/js/src/forum/components/LogInModal.tsx]: Sorry, was using this for testing! 🙈

**Comment by davwheat** [comment]
oops... that was the wrong git command...

have asked jasper for write access to his fork so i can resurrect this. have a backup of the code at https://github.com/flarum/framework/tree/jv/stackable-modals

**Comment by davwheat** [comment]
RE the scrolling bug: isn't this how the modals currently work in core?

https://user-images.githubusercontent.com/7406822/178990288-fede20f7-ccbc-4bf0-a72f-979446491969.mp4

**Comment by SychO9** [comment]
yea but that isn't how they work in this branch 🙈 `0:15` of my uploaded video shows behavior in this branch

**Comment by davwheat** [comment]
OH I didn't watch for long enough! Right, I'll get on that.

**Comment by davwheat** [comment]
@SychO9 Fixed, I hope.

**Comment by SychO9** [comment]
🙈 now you can't scroll inside a modal

File: framework/core/less/common/Modal.less
Comment (SychO9) [inline framework/core/less/common/Modal.less]: so the scrollbar shows up correctly but with `pointer-events: none` I can't scroll. But removing this also means you can't click outside to close 🤔

File: framework/core/js/src/common/components/ModalManager.tsx
Comment (SychO9) [inline framework/core/js/src/common/components/ModalManager.tsx:70]: Ideally, the backdrop would show behind the latest open modal and above the others. I think a simple CSS trick would do it: (other comment below for follow up)

```suggestion
            data-showing={!!this.attrs.state.modalList.length}
            style={{'--modal-count': this.attrs.state.modalList.length}}
```

File: framework/core/less/common/Modal.less
Comment (SychO9) [inline framework/core/less/common/Modal.less]:
```suggestion
  z-index: ~"calc(var(--zindex-modal) + var(--modal-count)) - 2";
```

**Review by SychO9 — COMMENTED** [review]
Using one backdrop seems great! I think only small tweaks are left.

**Comment by luceos** [comment]
The patience in this PR is admirable.

**Comment by davwheat** [comment]
PS: JS checks fail in some cases due to attempting to commit and push dist JS to Jasper's repo since the PR is based on the `main` branch.

**Comment by SychO9** [comment]
Under known issues: https://caniuse.com/pointer-events
> Moving the scrollbar on an object with pointer-events: none; works in Firefox, but doesn't work in either Chrome or IE.

What I did is introduce invisible backdrops under each modal so that we don't have to use `pointer-events` and be able to easily exit the modal while maintaining scrollabilty (*is that a word?*).

**Comment by davwheat** [comment]
Ahhhh that's a great find!

I didn't think of that idea and don't think I ever would have!

File: framework/core/js/src/common/components/ModalManager.tsx
Comment (SychO9) [inline framework/core/js/src/common/components/ModalManager.tsx]:
```suggestion
  protected handlePossibleBackdropClick(e: MouseEvent): void {
```

**Review by davwheat — APPROVED** [review]

**Comment by davwheat** [comment]
Rebased on `main`.

**Review by SychO9 — APPROVED** [review]

**Comment by davwheat** [comment]
phew that was a big one 😄

**Pull request merged**
